pkg/migrations/00001_init-schema.up.sql

```diff
@@ -1,3 +1,48 @@
-SELECT
-	1;
+-- Ensures that if the command-line node configuration mutates,
+-- the existing data in the DB is invalid
+CREATE TABLE node_info(
+	node_id INTEGER NOT NULL,
+	public_key BYTEA NOT NULL,
 
+	singleton_id SMALLINT PRIMARY KEY DEFAULT 1,
+	CONSTRAINT is_singleton CHECK (singleton_id = 1)
+);
+
+CREATE TABLE all_envelopes(
+	-- used to construct gateway_sid
+	id BIGSERIAL PRIMARY KEY,
+	originator_sid BIGINT NOT NULL,
+	topic BYTEA NOT NULL,
+	originator_envelope BYTEA NOT NULL
+);
+-- Client queries
+CREATE INDEX idx_all_envelopes_topic ON all_envelopes(topic);
+-- Node queries
+CREATE UNIQUE INDEX idx_all_envelopes_originator_sid ON all_envelopes(originator_sid);
+
+
+-- Process for originating envelopes:
+-- 1. Insert into originated_envelopes
+-- 2. Query staged_originated_envelopes, and for each envelope in order of ID:
+--   2.1. Construct and sign OriginatorEnvelope proto
+--   2.2. Atomically insert into all_envelopes and delete from originated_envelopes,
+--	      ignoring unique index violations on originator_sid
```

Comment:
A flaw with this approach is we will get a number of failures approximately proportionate to the rate of incoming publishes. Considering:
- Using pessimistic concurrency instead of optimistic concurrency
- Having a singleton worker handle this step

Reply:
Decided to use a singleton background task. Note that if multiple server instances are running, there will be one background task on each, therefore the amount of redundant work is proportional to the number of nodes. I assume this is fine, although it's theoretically possible to make only one task exist per cluster.

Another caveat is that the overall publish throughput of the cluster is constrained to the throughput of the biggest node, also assuming this is fine for now.